File: src/main/java/semicontinuity/idea/code/analyzer/graph/DAGraphViewRenderer2.java
package semicontinuity.idea.code.analyzer.graph;

import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Collectors;

import semicontinuity.idea.code.analyzer.graph.viewModel.Factory;

public class DAGraphViewRenderer2<N,
        NODE_PAYLOAD,
        COMP,
        IND_COMPS extends COMP,
        FANOUT extends COMP,
        DEP_COMPS extends COMP,
        NODE extends COMP> {

    private final Factory<NODE_PAYLOAD, COMP, IND_COMPS, FANOUT, DEP_COMPS, NODE> viewFactory;
    private final Function<N, NODE_PAYLOAD> payloadFunction;

    public DAGraphViewRenderer2(
            Factory<NODE_PAYLOAD, COMP, IND_COMPS, FANOUT, DEP_COMPS, NODE> viewFactory,
            Function<N, NODE_PAYLOAD> payloadFunction) {
        this.viewFactory = viewFactory;
        this.payloadFunction = payloadFunction;
    }


    public COMP render(DAGraph<N> graph) {
        if (!graph.hasEdges()) {
            var nodeViews = graph.nodes()
                    .stream()
                    .map(node -> viewFactory.newNode(payloadFunction.apply(node)))
                    .collect(Collectors.toList());
            return viewFactory.newIndependentComponents(nodeViews);
        }

        var decompose = graph.decompose();
        var components = decompose.entrySet()
                .stream()
                .map(this::renderIndependentComponent)
                .collect(Collectors.toList());

        if (components.size() == 1) {
            return components.get(0);
        } else {
            return viewFactory.newIndependentComponents(components);
        }
    }


    COMP renderIndependentComponent(Map.Entry<Set<N>, DAGraph<N>> rootsWithSubgraph) {
        var roots = rootsWithSubgraph.getKey();
        var subGraph = rootsWithSubgraph.getValue();
        var subGraphView = render(subGraph);

        var rootsViews = roots.stream()
                .map(r -> viewFactory.newNode(payloadFunction.apply(r)))
                .collect(Collectors.toList());

        if (rootsViews.size() == 1) {
            return viewFactory.newFanout(rootsViews.get(0), subGraphView);
        } else {
            return viewFactory.newDependentComponents(rootsViews, subGraphView);
        }
    }
}

File: src/main/java/semicontinuity/idea/code/analyzer/graph/DAGraphViewRenderer2.kt
package semicontinuity.idea.code.analyzer.graph

import semicontinuity.idea.code.analyzer.graph.viewModel.Factory
import java.util.function.Function

class DAGraphViewRenderer2<V, VERTEX_PAYLOAD, COMP, IND_COMPS : COMP, VERTEX : COMP, SPLIT : COMP, LAYER : COMP, SORT_KEY : Comparable<SORT_KEY>?, SPLIT0 : COMP>(
    viewFactory: Factory<VERTEX_PAYLOAD, COMP, IND_COMPS, VERTEX, SPLIT, LAYER, SPLIT0>,
    payloadFunction: Function<V, VERTEX_PAYLOAD>,
    sortKeyFunction: Function<V, SORT_KEY>,
) : DAGraphViewRenderer<V, VERTEX_PAYLOAD, COMP, IND_COMPS, VERTEX, SPLIT, LAYER, SORT_KEY, SPLIT0>(
    viewFactory,
    payloadFunction,
    sortKeyFunction
) {
    override fun doRenderGraphWithEdges(graph: DAGraph<V>): COMP =
        viewFactory.independentCompsOrFirst(
            DAGraphCutter(graph).cut().map { doRenderConnectedGraph(it) }
        )

    private fun doRenderConnectedGraph(graph: DAGraph<V>): COMP {
        val (front, back) = DAGraphSegregator(graph).segregate()

        return when {
            !front.hasVertices() -> doRenderNonEmptyGraph(back)
            !back.hasVertices() -> doRenderTreeGraph(front)
            else -> viewFactory.newSplit(
                doRenderTreeGraph(front),
                doRenderNonEmptyGraph(back)
            )
        }
    }

    private fun doRenderTreeGraph(graph: DAGraph<V>): COMP = doRenderGraphLayers(graph)
}
